include/ejabberd_c2s.hrl

```diff
@@ -29,6 +29,7 @@
                 tls_options = [],
                 tls_verify            :: verify_none | verify_peer,
                 authenticated = false :: authenticated_state(),
+                host_type = <<>>      :: binary(),
```

Comment:
Is there any reason to set an empty host type?

Reply:
otherwise, dialyzer complains :)

Reply:
Hm... maybe it has a reason to complain? What is the error? If it's in tests, then ok, but are you sure that we won't get an empty binary for an online user?

Reply:
it complains that `undefined` is not binary() type. and nope - we won't have it empty for an online user.
we set a valid host_type at the very early stage of session, actually authentication is impossible without it.

Reply:
could be `binary() | undefined` type

Reply:
Then it will complain about usage of `undefined` type everywhere where only binary() is expected.
There is no problem with the current type definition, there is no problem with the default value either.

Reply:
I like what @arcusfelis suggested. Dialyzer does not complain, I checked it.

Reply:
Dialyzer doesn't complain about it right now, because we don't really use that field yet. But I will start in the following PRs, I'm absolutely against this change. Dialyzer is not smart enough to understand FSM lifecycle and understand that in some state it cannot be undefined any more. 
empty binary is the same illegal value for the host type as undefined, we cannot configure it in toml.

Reply:
I still think Dialyzer would not complain because of how success typing works. However, it is a minor thing for me and I can merge the changes as they are now.

Reply:
did some extra tests, looks you're right. will change in the following PR.